### PR TITLE
Feature/tasks

### PR DESCRIPTION
**Pull request by anrgl**
- tasks

### CODE REVIEW

File: app/javascript/components/ColumnHeader/ColumnHeader.js [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+    </div>
+  );
+};
+
+ColumnHeader.propTypes = {
+  onLoadMore: PropTypes.func.isRequired,
+  column: PropTypes.shape({
+    id: PropTypes.string.isRequired,
+    title: PropTypes.string.isRequired,
+    cards: PropTypes.array.isRequired,
```
Comment (dvislov) [inline app/javascript/components/ColumnHeader/ColumnHeader.js]: Need to describe array's elements

File: app/javascript/components/Form/Form.js [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+import { has } from "ramda";
+
+import TextField from "@material-ui/core/TextField";
+
+import useStyles from "./useStyles";
+
+const Form = ({ errors, onChange, task }) => {
+  const handleChangeTextField = (fieldName) => (event) =>
+    onChange({ ...task, [fieldName]: event.target.value });
+  const styles = useStyles();
```
Comment (dvislov) [inline app/javascript/components/Form/Form.js]: Hooks should be at the beginning of the component

File: app/javascript/utils/fetchHelper.js [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+    const body = decamelize(json);
+    return axios.post(url, body).then(camelize);
+  },
+
+  put(url, json) {
+    const body = decamelize(json);
+    return axios.put(url, body).then(camelize);
+  },
+
+  delete(url, _json) {
```
Comment (dvislov) [inline app/javascript/utils/fetchHelper.js]: `_json` is unused

File: app/javascript/repositories/TasksRepository.js [hunk 4]
```diff
@@ -0,0 +1,29 @@
+import routes from "../routes";
```
Comment (dvislov) [inline app/javascript/repositories/TasksRepository.js]: Avoid relative paths

File: app/javascript/forms/EditPopup.js [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+import Modal from "@material-ui/core/Modal";
+import Card from "@material-ui/core/Card";
+import CardHeader from "@material-ui/core/CardHeader";
+import CardContent from "@material-ui/core/CardContent";
+import CardActions from "@material-ui/core/CardActions";
+import { CircularProgress } from "@material-ui/core";
+import Button from "@material-ui/core/Button";
+import CloseIcon from "@material-ui/icons/Close";
+
+import Form from "../components/Form";
```
Comment (dvislov) [inline app/javascript/forms/EditPopup.js]: Avoid relative paths

File: app/javascript/forms/EditPopup.js [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+  onLoadCard,
+  onCardUpdate,
+}) => {
+  const [task, setTask] = useState(null);
+  const [isSaving, setSaving] = useState(false);
+  const [errors, setErrors] = useState({});
+  const styles = useStyles();
+
+  useEffect(() => {
+    onLoadCard(cardId).then(setTask);
```
Comment (dvislov) [inline app/javascript/forms/EditPopup.js:25]: Let's rewrite to `async/await`

File: app/javascript/forms/EditPopup.js [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+              : `Task # ${task.id} [${task.name}]`
+          }
+        />
+        <CardContent>
+          {isLoading ? (
+            <div className={styles.loader}>
+              <CircularProgress />
+            </div>
+          ) : (
+            <Form errors={errors} onChange={setTask} task={task} />
```
Comment (dvislov) [inline app/javascript/forms/EditPopup.js:73]: Better not to pass state change function as is down the components tree, create the handler like `handleFormChange` and call `setTask` inside

File: app/javascript/forms/AddPopup.js [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+      setErrors(error || {});
+
+      if (error instanceof Error) {
+        alert(`Creation Failed! Error: ${error.message}`);
+      }
+    });
+  };
+  const handleChangeTextField = (fieldName) => (event) =>
+    changeTask({ ...task, [fieldName]: event.target.value });
+  const styles = useStyles();
```
Comment (dvislov) [inline app/javascript/forms/AddPopup.js]: Move hook upper

File: app/javascript/forms/AddPopup.js [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+    onCreateCard(task).catch((error) => {
+      setSaving(false);
+      setErrors(error || {});
+
+      if (error instanceof Error) {
+        alert(`Creation Failed! Error: ${error.message}`);
+      }
+    });
+  };
+  const handleChangeTextField = (fieldName) => (event) =>
```
Comment (dvislov) [inline app/javascript/forms/AddPopup.js:37]: https://gitlab.com/dualbootpartners/docs/general/-/wikis/en/Practice%20Specific%20Principles/Js%20Frontend%20Architecture#64-name-your-event-handlers-and-pass-them-as-props-in-a-good-way

File: app/javascript/forms/AddPopup.js [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+import CloseIcon from "@material-ui/icons/Close";
+import IconButton from "@material-ui/core/IconButton";
+import Modal from "@material-ui/core/Modal";
+import TextField from "@material-ui/core/TextField";
+
+import TaskForm from "forms/TaskForm";
+
+import useStyles from "./useStyles";
+
+const AddPopup = ({ onClose, onCreateCard }) => {
```
Comment (dvislov) [inline app/javascript/forms/AddPopup.js]: https://gitlab.com/dualbootpartners/docs/general/-/wikis/en/Practice%20Specific%20Principles/Js%20Frontend%20Architecture#61-properties-extracting-rules, check everywhere in the request

File: app/javascript/forms/AddPopup.js [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+  const [task, changeTask] = useState(TaskForm.defaultAttributes());
```
Comment (dvislov) [inline app/javascript/forms/AddPopup.js]: Do not invent custom names for setState methods, should be start with `set`: `[task, setTask]`

File: app/javascript/forms/AddPopup.js [hunk 12]
```diff
@@ -0,0 +1,93 @@
...
+          </Button>
+        </CardActions>
+      </Card>
+    </Modal>
+  );
+};
+
+AddPopup.propTypes = {
+  onClose: PropTypes.func.isRequired,
+  onCreateCard: PropTypes.func.isRequired,
```
Comment (dvislov) [inline app/javascript/forms/AddPopup.js:91]: https://gitlab.com/dualbootpartners/docs/general/-/wikis/en/Practice%20Specific%20Principles/Js%20Frontend%20Architecture#64-name-your-event-handlers-and-pass-them-as-props-in-a-good-way

File: app/javascript/components/Task/Task.js [hunk 13]
```diff
@@ -0,0 +1,39 @@
...
+import Card from "@material-ui/core/Card";
+import CardHeader from "@material-ui/core/CardHeader";
+import CardContent from "@material-ui/core/CardContent";
+import Typography from "@material-ui/core/Typography";
+import IconButton from "@material-ui/core/IconButton";
+import EditIcon from "@material-ui/icons/Edit";
+
+import useStyles from "./useStyle";
+
+const Task = ({ task, onClick }) => {
```
Comment (dvislov) [inline app/javascript/components/Task/Task.js]: Don't see `onClick` in the PropTypes, and it is not a good name (very abstract)

File: app/javascript/components/TaskBoard/TaskBoard.js [hunk 14]
```diff
@@ -1,39 +1,205 @@
...
+  const handleOpenAddPopup = () => {
+    setMode(MODES.ADD);
+  };
+
+  const handleClose = () => {
+    setMode(MODES.NONE);
+    setOpenedTaskId(null);
+  };
+
+  const [mode, setMode] = useState(MODES.NONE);
```
Comment (dvislov) [inline app/javascript/components/TaskBoard/TaskBoard.js]: move to the top

File: app/javascript/components/ColumnHeader/ColumnHeader.js [hunk 15]
```diff
@@ -0,0 +1,51 @@
+import React from "react";
+import PropTypes from "prop-types";
+
+import IconButton from "@material-ui/core/IconButton";
+import SystemUpdateAltIcon from "@material-ui/icons/SystemUpdateAlt";
+
+import useStyles from "./useStyles";
+
+const ColumnHeader = ({ column, onLoadMore }) => {
```
Comment (dvislov) [inline app/javascript/components/ColumnHeader/ColumnHeader.js]: https://gitlab.com/dualbootpartners/docs/general/-/wikis/en/Practice%20Specific%20Principles/Js%20Frontend%20Architecture#61-properties-extracting-rules, check in all the components in the request

File: app/javascript/components/Form/Form.js [hunk 16]
```diff
@@ -0,0 +1,54 @@
+import React from "react";
+import PropTypes from "prop-types";
+import { has } from "ramda";
+
+import TextField from "@material-ui/core/TextField";
+
+import useStyles from "./useStyles";
+
+const Form = ({ errors, onChange, task }) => {
+  const handleChangeTextField = (fieldName) => (event) =>
```
Comment (dvislov) [inline app/javascript/components/Form/Form.js:12]: https://gitlab.com/dualbootpartners/docs/general/-/wikis/en/Practice%20Specific%20Principles/Js%20Frontend%20Architecture#64-name-your-event-handlers-and-pass-them-as-props-in-a-good-way

File: app/javascript/components/TaskBoard/TaskBoard.js [hunk 17]
```diff
@@ -1,39 +1,205 @@
...
+
+import Task from "components/Task";
+import ColumnHeader from "../ColumnHeader";
+import TasksRepository from "repositories/TasksRepository";
+import { Fab } from "@material-ui/core";
+import AddIcon from "@material-ui/icons/Add";
+import useStyles from "./useStyles";
+import AddPopup from "../../forms/AddPopup";
+import EditPopup from "../../forms/EditPopup";
+import TaskForm from "../../forms/TaskForm";
```
Comment (dvislov) [inline app/javascript/components/TaskBoard/TaskBoard.js]: avoid relative paths for these three lines

File: app/javascript/components/TaskBoard/TaskBoard.js [hunk 18]
```diff
@@ -1,39 +1,205 @@
...
-      ]
-    }
-  ]
+import React, { useEffect, useState } from "react";
+import KanbanBoard from "@lourenci/react-kanban";
+import "@lourenci/react-kanban/dist/styles.css";
+import { propOr } from "ramda";
+
+import Task from "components/Task";
+import ColumnHeader from "../ColumnHeader";
```
Comment (dvislov) [inline app/javascript/components/TaskBoard/TaskBoard.js]: Relative path is not ok, or maybe ColumnHeader should be inside TaskBoard/components?

File: app/javascript/components/TaskBoard/TaskBoard.js [hunk 19]
```diff
@@ -1,39 +1,205 @@
...
+      .then(() => {
+        loadColumnInitial(destination.toColumnId);
+        loadColumnInitial(source.fromColumnId);
+      })
+      .catch((error) => {
+        alert(`Move failed! ${error.message}`);
+      });
+  };
+
+  const MODES = {
```
Comment (dvislov) [inline app/javascript/components/TaskBoard/TaskBoard.js]: No need to create this const every render, can be moved outside the component

File: app/javascript/forms/AddPopup.js [hunk 20]
```diff
@@ -0,0 +1,93 @@
+import React, { useState } from "react";
```
Comment (dvislov) [inline app/javascript/forms/AddPopup.js:1]: This file should be not in `forms` folder

File: app/controllers/api/v1/tasks_controller.rb [hunk 21]
```diff
@@ -1,12 +1,13 @@
 class Api::V1::TasksController < Api::V1::ApplicationController
   def index
-    tasks = Task.all
-                .ransack(ransack_params)
-                .result
-                .page(page)
-                .per(per_page)
-
-    respond_with(tasks, each_serializer: TaskSerializer, root: 'items', meta: build_meta(tasks))
+    tasks = Task.all.
```
Comment (YJorge) [inline app/controllers/api/v1/tasks_controller.rb:3]: странное форматирование.  Надо исправить

File: app/javascript/components/Task/Task.js [hunk 22]
```diff
@@ -0,0 +1,40 @@
...
+import CardContent from "@material-ui/core/CardContent";
+import Typography from "@material-ui/core/Typography";
+import IconButton from "@material-ui/core/IconButton";
+import EditIcon from "@material-ui/icons/Edit";
+
+import useStyles from "./useStyle";
+
+const Task = (props) => {
+  const { task, onClick } = props;
+  const styles = useStyles;
```
Comment (YJorge) [inline app/javascript/components/Task/Task.js:15]: это точно работает?

File: app/javascript/forms/EditPopup.js [hunk 23]
```diff
@@ -0,0 +1,109 @@
+import React, { useEffect, useState } from "react";
```
Comment (YJorge) [inline app/javascript/forms/EditPopup.js:1]: этот файл не должен быть в папке forms

**Pull request merged**